### ISSUE TITLE
request for a new [monitor~] object

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1937]
hey, here's a fun idea, an object that outputs all that is being sent to a [dac~] channel, for monitoring.

I know we could include many [throw~] objects for a bus and then monitor it, so this may likely get "hmmmm, sound more like a convenience external idea", but maybe it's worth giving it a shot?

The thing is that in order for that to work I'd need to also design an external just to replace [dac~] and have each output channel send to an internal "[throw~]" and that feels like too much...

anyway, just throwing this one in...

cheers

**Comment by umlaeute** [comment]
> The thing is that in order for that to work I'd need to also design an external just to replace [dac~] and have each output channel send to an internal "[throw~]" and that feels like too much...

so you just use a `[DAC~]` **abstraction** instead (and the abstraction has the `[throw~]`s).
it also makes it somewhat less recursive if you want to send the output of the `[monitor~]` to the dac for monitoring.

**Comment by porres** [comment]
> so you just use a [DAC~] abstraction instead (and the abstraction has the [throw~]s).

Yeah, that would an external, and by being 'too much' I'm saying I don't like the idea of having a separate [dac~]. And the main point would be to check levels of whatever is going out to the sound card, not via this or that object... this is what makes me think there's no way to do this with an external.

> it also makes it somewhat less recursive if you want to send the output of the [monitor~] to the dac for monitoring.

Not sure what you mean here. I guess the main usage of this idea would be checking [vu] GUIs so one can see if output levels are alright. I guess one can redirect monitor input from what's goin out to [dac~ 1 2] to another stereo monitor output, like [dac~ 3 4]. Haven't thought of that and seems like an interesting use case! 

> Although doing it in an external wouldn't be too bad either at the cost of an additional array per channel. (per pd instance?)

I'd be glad if I could design this as an external, but I can't see how to do this unless I fork Pd and modify [dac~], do you mean I can do it? Maybe the core can be changed a bit to allow us to do crazy things like this with an external?

**Comment by Ant1r** [comment]
Really, you can already do everything that has been mentioned here and more (except of course the actual final output level in the Test Audio and Midi patch), if you consistently use a dac replacement abstraction.

If you [catch~] yourself the audio outputs of your patches into an audio master patch, you can have a master level control, master vu, master limiter and EQ and so on.

So that seems overkill to me to add a [monitor ~] object that would only give you the vu, without allow you to e.g adjust the level.

**Comment by porres** [comment]
> Really, you can already do everything that has been mentioned here and more (except of course the actual final output level in the Test Audio and Midi patch), if you consistently use a dac replacement abstraction.

yup, I know I can, I mentioned in the begging and the last message, but also brought this out level example as something that actually motivated me for this feature, which is unique.

Not sure about adding a mixer and all, seems like something one should build and more of a higher level thing, or something to apply before reaching [dac~].

This idea also relates to how MAX (and also PlugData) have a [vu] that displays a master output and how you can also control this master level. Pd Extended used to have a meter display at the Pd window (no master level though). Something like MAX/Plugdata would maybe be nice, more kind of what you wanted, but not this is what I'm after, I'm hoping for a lower level thing. I wouldn't complaing of course but I see this would also require some GUI change as in MAX/PlugData so we'd always have this meter and gain control somewhere.

I get it might be a bit too much to add kinda of a special object to the mix that is something we all kinda lived without, my problem was how to do it with an external (and I'm not sure there is a way yet, I didn't get how I could do it if there is). What just came up to me is that I can in fact design a modified version of [dac~] as an external that would replace the original one. Of course it'd be backwards compatible but it would also allow me to pair it up with this other external and maybe go crazy with other ideas...

**Comment by sebshader** [comment]
I think the main concern is the feedback issue where the output from the monitor goes back into the ```[dac~]```. Those other cases from pd-extended don't have that issue.
But I suppose the monitor would always lag by 1 block anyways

**Comment by porres** [comment]
What's not clear to me is if there's a way to write an external that reads and outputs audio from whatever is coming out of Pd without the need to worry about modifying [dac~]... if so, this kinda settles this issue as I can do it as an external. Also, if so, how? Could anyone help me or should I ask pd-dev?

thanks

**Comment by porres** [comment]
well, like I said, there could be an external also that modifies [dac~] in the worst case scenario, but honestly, I'm not sure how to do that either and include some sort of built-in "throw~" in it so it communicates with a built in "catch~" in an external...

And now for another idea...

What if [dac~] has outlets? The outlets could output the summed bus of whatever is coming into it from all [dac~] instances in Pd. This would be a "monitor" output and wouldn't require a new object at all...

The monitor output could also not be the default setting. Like, a "-m" flag (for monitoring) could be given to make it create monitoring outlets. The output should also not even lag by any audio block ;) 

seems like a good idea, I just wouldn't also know how to do it, or try it as a prototype in a branch or as an external.

**Comment by Spacechild1** [comment]
> What's not clear to me is if there's a way to write an external that reads and outputs audio from whatever is coming out of Pd without the need to worry about modifying [dac~]...

See https://github.com/pure-data/pure-data/issues/1937#issuecomment-1487256748

> What if [dac~] has outlets? The outlets could output the summed bus of whatever is coming into it from all [dac~] instances in Pd. This would be a "monitor" output and wouldn't require a new object at all...

You would only get the complete sum at the end of the DSP graph. Unless you add an additional buffer that keeps the audio around. But then you might still get a one block delay between different source, depending on where they are in the graph. Therefore I don't think that outlets for `[dac~]` would make sense.

I agree with the suggestion to just use `[throw~]` instead.

**Comment by porres** [comment]
> See https://github.com/pure-data/pure-data/issues/1937#issuecomment-1487256748

If I can make an external by reading from read from "STUFF->st_soundout", then I'm only missing "how"

> You would only get the complete sum in the last [dac~] object. Therefore I don't think that outlets for [dac~] would make much sense.

hmmmm :/

what if those actually read from "STUFF->st_soundout" then with a delayed block of audio?

**Comment by Spacechild1** [comment]
> what if those actually read from "STUFF->st_soundout" then with a delayed block of audio?

`st_soundout` must be cleared before each scheduler tick. However, Pd *could* keep a second copy around, so objects can read it (albeit with a one block delay).

In general, this whole problem is very specific to Pd (and Max) because we have little control over the graph order. (We must resort to tricks like subpatches with dummy signal connections.) In SuperCollider, for example, this is not an issue at all because Synths write to well-defined bus numbers and you can always add a new Node at the end of the graph to do any kind of post-processing, monitoring, etc.
